hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferProxy.java

```diff
@@ -0,0 +1,213 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.ringbuffer.impl;
+
+import com.hazelcast.config.RingbufferConfig;
+import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.core.IFunction;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.ringbuffer.OverflowPolicy;
+import com.hazelcast.ringbuffer.ReadResultSet;
+import com.hazelcast.ringbuffer.Ringbuffer;
+import com.hazelcast.ringbuffer.impl.operations.AddAllOperation;
+import com.hazelcast.ringbuffer.impl.operations.AddOperation;
+import com.hazelcast.ringbuffer.impl.operations.GenericOperation;
+import com.hazelcast.ringbuffer.impl.operations.ReadManyOperation;
+import com.hazelcast.ringbuffer.impl.operations.ReadOneOperation;
+import com.hazelcast.spi.AbstractDistributedObject;
+import com.hazelcast.spi.InternalCompletableFuture;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.OperationService;
+
+import java.util.Collection;
+
+import static com.hazelcast.ringbuffer.OverflowPolicy.OVERWRITE;
+import static com.hazelcast.ringbuffer.impl.RingbufferService.SERVICE_NAME;
+import static com.hazelcast.ringbuffer.impl.operations.GenericOperation.OPERATION_HEAD;
+import static com.hazelcast.ringbuffer.impl.operations.GenericOperation.OPERATION_REMAINING_CAPACITY;
+import static com.hazelcast.ringbuffer.impl.operations.GenericOperation.OPERATION_SIZE;
+import static com.hazelcast.ringbuffer.impl.operations.GenericOperation.OPERATION_TAIL;
+import static com.hazelcast.util.ExceptionUtil.rethrowAllowInterrupted;
+import static com.hazelcast.util.Preconditions.checkFalse;
+import static com.hazelcast.util.Preconditions.checkNotNegative;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static com.hazelcast.util.Preconditions.checkTrue;
+import static java.lang.String.format;
+
+/**
+ * The serverside proxy to access a {@link Ringbuffer}.
+ *
+ * @param <E> the type of the elements in the ringbuffer.
+ */
+public class RingbufferProxy<E> extends AbstractDistributedObject<RingbufferService> implements Ringbuffer<E> {
+
+    // to prevent overloading on readMany/addAll, there is a hard limit
+    static final int MAX_BATCH_SIZE = 1000;
+
+    private final String name;
+    private final int partitionId;
+    private final RingbufferConfig config;
+
+    public RingbufferProxy(NodeEngine nodeEngine, RingbufferService service, String name) {
+        super(nodeEngine, service);
+        this.name = name;
+        this.partitionId = nodeEngine.getPartitionService().getPartitionId(name);
```

Comment:
To be able to handle partition-aware naming, you should use `StringPartitioningStrategy` here. For example in `QueueProxySupport` this is done with:
```java
this.partitionId = nodeEngine.getPartitionService().getPartitionId(getNameAsPartitionAwareData());
```

`AbstractDistributedObject .getNameAsPartitionAwareData()` is:
```java
protected Data getNameAsPartitionAwareData() {
     String name = getName();
     return getNodeEngine().getSerializationService().toData(name, PARTITIONING_STRATEGY);
}
```